devito/ir/cgen/printer.py

```diff
@@ -33,25 +33,101 @@ class CodePrinter(C99CodePrinter):
         Options for code printing.
     """
     _default_settings = {'compiler': None, 'dtype': np.float32,
-                         **C99CodePrinter._default_settings}
+                         **CodePrinter._default_settings}
+
+    _func_prefix = {}
+    _func_litterals = {}
+
+    _qualifiers_mapper = {
+        'is_const': 'const',
+        'is_volatile': 'volatile',
+        '_mem_constant': 'static',
+        '_mem_shared': '',
+    }
+
+    _prec_litterals = {np.float32: 'F', np.complex64: 'F'}
+
+    _restrict_keyword = 'restrict'
+
+    _includes = []
+
+    _namespaces = []
+
+    _headers = [('_POSIX_C_SOURCE', '200809L')]
 
     @property
     def dtype(self):
-        return self._settings['dtype']
+        try:
+            return self._settings['dtype'].nptype
+        except AttributeError:
+            return self._settings['dtype']
 
     @property
     def compiler(self):
-        return self._settings['compiler']
+        return self._settings['compiler'] or configuration['compiler']
+
+    def doprint(self, expr, assign_to=None):
+        """
+        The sympy code printer does a lot of extra we do not need as we handle all of
+        it in the compiler so we directly defaults to `_print`.
+        """
+        return self._print(expr)
 
-    def single_prec(self, expr=None):
-        dtype = sympy_dtype(expr) if expr is not None else self.dtype
-        return dtype in [np.float32, np.float16]
+    def _prec(self, expr):
+        try:
+            dtype = sympy_dtype(expr, default=self.dtype)
+        except TypeError:
+            return self.dtype
```

Comment:
imho this is kinda weird; you're passing a `default=...`, so you would expect the callee, `sympy_dtype`, to handle such a `TypeError` and just `return default`... which would avoid a try-except here at the call site